File: 0100_Same_Tree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// Recursion T = O(n), S = O(n)
class Solution {
  public:
    bool dfs(TreeNode *p, TreeNode *q) {
        if (p == nullptr && q == nullptr) {
            return true;
        }
        if (p == nullptr || q == nullptr) {
            return false;
        }
        return p->val == q->val && dfs(p->left, q->left) &&
               dfs(p->right, q->right);
    }
    bool isSameTree(TreeNode *p, TreeNode *q) { return dfs(p, q); }
};

// Iteration T = O(n), S = O(n)
class Solution {
  public:
    bool isSameTree(TreeNode *p, TreeNode *q) {
        stack<TreeNode *> st;
        st.push(p);
        st.push(q);
        while (!st.empty()) {
            TreeNode *s = st.top();
            st.pop();
            TreeNode *t = st.top();
            st.pop();
            if (s == nullptr && t == nullptr) {
                continue;
            }
            if (s == nullptr || t == nullptr) {
                return false;
            }
            if (s->val != t->val) {
                return false;
            }
            st.push(s->left);
            st.push(t->left);
            st.push(s->right);
            st.push(t->right);
        }
        return true;
    }
};

File: 0100_Same_Tree/golang/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// recursion
func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p == nil || q == nil {
		return p == nil && q == nil
	}
	if p.Val != q.Val {
		return false
	}
	return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

// iteration
func isSameTree(p *TreeNode, q *TreeNode) bool {
	st := []*TreeNode{}
	st = append(st, p, q)
	for len(st) != 0 {
		p = st[len(st)-1]
		st = st[:len(st)-1]
		q = st[len(st)-1]
		st = st[:len(st)-1]
		if p == nil && q == nil {
			continue
		}
		if p == nil || q == nil || p.Val != q.Val {
			return false
		}
		st = append(st, p.Left, q.Left, p.Right, q.Right)
	}
	return true
}